R/dt_spanners.R

```diff
@@ -87,17 +113,143 @@ dt_spanners_print <- function(data, include_hidden = TRUE, ids = FALSE) {
   } else {
     vars <- dt_boxhead_get_vars(data = data)
   }
-
-  vars_list <- rep(NA_character_, length(vars)) %>% magrittr::set_names(vars)
+  vars_vec <- rep(NA_character_, length(vars))
+  names(vars_vec) <- vars
 
   for (i in seq_len(nrow(spanners))) {
 
     if (ids) {
-      vars_list[spanners$vars[[i]]] <- spanners$spanner_id[i]
+      vars_vec[spanners$vars[[i]]] <- spanners$spanner_id[i]
+    } else {
+      vars_vec[spanners$vars[[i]]] <- spanners$built[[i]]
+    }
+  }
+
+  unname(vars_vec[names(vars_vec) %in% vars])
+}
+
+dt_spanners_print_matrix <- function(
+    data,
+    include_hidden = TRUE,
+    ids = FALSE,
+    omit_columns_row = FALSE
+) {
+
+  spanners_tbl <- dt_spanners_get(data = data)
+
+  remove_entry <- rep(FALSE, nrow(spanners_tbl))
+
+  if (!include_hidden) {
+    vars <- dt_boxhead_get_vars_default(data = data)
+  } else {
+    vars <- dt_boxhead_get_vars(data = data)
+  }
+
+  # Keep only the columns in `spanners_tbl` that are in the `vars`;
+  # this will exclude spanners if their columns aren't in `vars` (this
+  # is possible when hiding columns via `col_hide()`
+  for (i in seq_len(nrow(spanners_tbl))) {
+
+    vars_spanner_i <- unlist(spanners_tbl[i, ][["vars"]])
+    vars_spanner_i_flt <- base::intersect(vars_spanner_i, vars)
+
+    if (length(vars_spanner_i_flt) < 1) {
+      # If none of the columns under the proposed spanner are
+      # set as visible columns, then we need to mark the spanner
+      # entry for deletion from `spanners_tbl`
+      remove_entry[i] <- TRUE
     } else {
-      vars_list[spanners$vars[[i]]] <- spanners$built[[i]]
+      spanners_tbl[i, ][["vars"]] <- list(vars_spanner_i_flt)
+    }
+  }
+
+  # Remove entries from `spanners_tbl` as needed
+  spanners_tbl <- spanners_tbl[!remove_entry, ]
+
+  # Get the maximum height of the spanner matrix; this won't
+  # necessarily be the finalized height since we will collapse
+  # empty rows
+  if (nrow(spanners_tbl) < 1) {
+    spanner_height <- 0
+  } else {
+    spanner_height <- max(spanners_tbl[["spanner_level"]])
+  }
+
+  vars_vec <- rep(NA_character_, length(vars))
+  names(vars_vec) <- vars
+
+  # Initialize matrix to serve as boxhead representation
+  columns_mat <- matrix(vars, nrow = 1, ncol = length(vars))
+
+  columns_mat <-
+    rbind(
+      columns_mat,
+      if (spanner_height > 0) {
+      matrix(vars_vec, nrow = spanner_height, ncol = length(vars_vec))
+      }
+    )
+  colnames(columns_mat) <- vars
+
+  spanners_tbl <- spanners_tbl %>% dplyr::filter(spanner_level >= 1)
+
+  columns_mat_id <- columns_mat_labels <- columns_mat
+
+  if (spanner_height > 0) {
+
+    for (i in seq_len(nrow(spanners_tbl))) {
+
+      columns_mat_id[spanners_tbl$spanner_level[[i]] + 1, spanners_tbl$vars[[i]]] <-
```

Comment:
Consider having columns_mat not contain the vars until the last minute--so it's not necessary to have `+ 1`.

Reply:
Now changed in that way.